src/Connector.sol

```diff
@@ -129,71 +164,122 @@ contract CentrifugeConnector {
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        uint128 currency = currencyAddressToId[currencyAddress];
+        require(currency != 0, "CentrifugeConnector/unknown-currency");
+        require(poolCurrencies[poolId][currencyAddress], "CentrifugeConnector/pool-currency-not-allowed");
+
+        gateway.decreaseRedeemOrder(poolId, trancheId, msg.sender, currency, amount);
     }
 
     function collectRedeem(uint64 poolId, bytes16 trancheId) public {
-        // TODO(nuno)
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        gateway.collectRedeem(poolId, trancheId, address(msg.sender));
     }
 
-    function collectForRedeem(uint64 poolId, bytes16 trancheId, bytes32 userAddress) public {
-        // TODO(nuno)
+    function collectForRedeem(uint64 poolId, bytes16 trancheId, bytes32 recipient) public {
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        gateway.collectForRedeem(poolId, trancheId, address(msg.sender), recipient);
     }
 
     function collectInvest(uint64 poolId, bytes16 trancheId) public {
-        // TODO(nuno)
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        gateway.collectInvest(poolId, trancheId, address(msg.sender));
     }
 
-    function collectForInvest(uint64 poolId, bytes16 trancheId, bytes32 userAddress) public {
-        // TODO(nuno)
+    function collectForInvest(uint64 poolId, bytes16 trancheId, bytes32 recipient) public {
+        RestrictedTokenLike token = RestrictedTokenLike(tranches[poolId][trancheId].token);
+        require(address(token) != address(0), "CentrifugeConnector/unknown-tranche-token");
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        gateway.collectForInvest(poolId, trancheId, address(msg.sender), recipient);
     }
 
     // --- Incoming message handling ---
-    // todo(nuno): store currency and decimals
-    function addPool(uint64 poolId, uint128 currency, uint8 decimals) public onlyGateway {
+    function addCurrency(uint128 currency, address currencyAddress) public onlyGateway {
+        currencyIdToAddress[currency] = currencyAddress;
```

Comment:
I think in the sync last week, we came to the conclusion that this should not be allowed for security reasons, and this should be immutable.

Reply:
Thanks, done. I think I found an edge case tho:

    // addCurrency(currency 1, address 0x123) -> Works
    // addCurrency(currency 1, address 0x456) -> Fails because there's already an address set for currency 1
    // addCurrency(currency 0, 0x123) -> Works because there's no address set for currency 0, even tho this same address is set for currency 1


In addition to `require(currencyIdToAddress[currency] == address(0),`, we should also check `require(currencyAddressToId[currencyAddress] == 0)`, to ensure that the same address doesn't get mapped to multiple currency ids. 

Or do we want to delegate this concern to Centrifuge? I.e, the mapping `Id <> EVM address` in Centrifuge should ensures that it's 1:1; This becomes an issue if we need to fix the `evm` address tho, since that requires creating a new mapping on Centrifuge 🤔

Reply:
I think you're right that we should also add this check. What this would block, is ever re-using a currency address as a different ID; I'm trying to think through if there's any edge case where this might be required, but I can't think of anything.

> This becomes an issue if we need to fix the evm address tho, since that requires creating a new mapping on Centrifuge

I think this is already required by the first `require`, and will be the right path to fix this, yeah. Which means we should just do everything we can to avoid incorrect ID <> EVM address mappings :)

Reply:
Done ✅